ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java

```diff
@@ -237,47 +216,55 @@ protected ScheduledExecutorService executor() {
      * @param windowEnd the end time in ms of the current window
      */
     private void processHeartbeats(final long windowStart, final long windowEnd) {
+
+      final Map<KsqlHost, HostStatus> copyOnWrite = new HashMap<>(hostsStatus.get());
       // No heartbeats received -> mark all hosts as dead
       if (receivedHeartbeats.isEmpty()) {
-        hostsStatus.replaceAll((host, status) -> {
-          if (!host.equals(localHostString)) {
-            return status.copyWithStatus(false);
+        copyOnWrite.replaceAll((host, status) -> {
+          if (!host.equals(localHost)) {
+            return status.withHostAlive(false);
           }
           return status;
         });
+        hostsStatus.set(ImmutableMap.copyOf(copyOnWrite));
```

Comment:
If `DiscoverClusterService` runs and discovers a new host while this block is running, we will overwrite whatever changes happened there. This might be okay, as eventually the `DisocverClusterService` will "rediscover" whatever was overwritten, but this is something to point out - at a minimum we should comment acknowledging that, at best we may want to synchronize the two. Same goes for L258

more concerning is perhaps the other way - a concurrent `DiscoverClusterService#runOneIteration` will overwrite the processheartbeats result